coffee_maker_navigator_2/lib/di/di/dependency_container_manager.dart

```diff
@@ -0,0 +1,137 @@
+import 'package:coffee_maker_navigator_2/di/di/dependency_containers/dependency_container.dart';
+
+/// The `DependencyContainerManager` is a singleton class responsible for managing
+/// the lifecycle of dependency containers in the application.
+///
+/// It initializes app-level dependencies at startup and manages feature-level
+/// dependencies dynamically, creating and disposing them based on their usage.
+class DependencyContainerManager {
+  static final _instance = DependencyContainerManager._internal();
+
+  // App-level dependencies that live for the entire duration of the app.
+  final AppLevelDependencyContainer _appLevelDependencies =
+      AppLevelDependencyContainer();
+
+  /// A map that registers container factories by their type.
+  ///
+  /// This map holds factory functions that are responsible for creating instances
+  /// of dependency containers. Each entry in the map associates a type [T] with a
+  /// factory function that produces a [DependencyContainer] of that type.
+  ///
+  /// The key is the type of the dependency container, and the value is a function
+  /// that takes an existing [AppLevelDependencies] and returns a new instance of a
+  /// [DependencyContainer] for the specified type [T].
+  final _containerFactories =
+      <Type, DependencyContainer Function(AppLevelDependencyContainer)>{};
+
+  // A map that tracks the subscribers for each container type.
+  final _containerSubscribers = <Type, Set<Object>>{};
+
+  // A map that holds the currently active containers by their type.
+  // An active container is one that is currently in use, meaning it has at least one subscriber.
+  final _activeContainers = <Type, DependencyContainer>{};
+
+  /// Returns the single instance of [DependencyContainerManager].
+  static DependencyContainerManager get instance => _instance;
+
+  /// Factory constructor to return the singleton instance.
+  factory DependencyContainerManager() => _instance;
+
+  /// Internal constructor for singleton implementation.
+  DependencyContainerManager._internal();
+
+  /// Initializes the app-level dependencies.
+  Future<void> init() async {
+    await _appLevelDependencies.init();
+    registerContainerFactory<AppLevelDependencyContainer>(
+        (_) => _appLevelDependencies);
+    subscribeToContainer<AppLevelDependencyContainer>(this);
+  }
+
+  /// Registers a dependency container factory for a specific type [T].
+  ///
+  /// This method allows you to register a factory function that will be responsible
+  /// for creating instances of the dependency container for the specified dependency container
+  /// type [T].
+  ///
+  /// [factory]: A function that takes an existing [DependencyContainer] and returns
+  /// a new instance of a [DependencyContainer] for the specified type [T].
+  void registerContainerFactory<T>(
+      DependencyContainer Function(DependencyContainer) factory) {
```

Comment:
Mismatching the signature of functions with _containerFactories, maybe to make it easy to use a typeDef?

Reply:
f98e03b04bc09617b1c085bd9f662fab7aa52ed9 done